upgrade_downgrade/common.sh

```diff
@@ -0,0 +1,186 @@
+      return 0
+  fi
+  echo "Response from server for kubectl get: "
+  echo "${resp}"
+  return 1
+}
+
+deleteWithWait() {
+  # Don't complain if resource is already deleted.
+  echo_and_run_quiet kubectl delete "${1}" -n "${3}" "${2}"
+  withRetries 60 10 checkIfDeleted "${1}" "${2}" "${3}"
+}
+
+_waitForPodsReady() {
+  pods_str=$(kubectl -n "${1}" get pods | tail -n +2 )
+  arr=()
+  while read -r line; do
+    arr+=("$line")
+  done <<< "$pods_str"
+
+  ready="true"
+  for line in "${arr[@]}"; do
+    if [[ ${line} != *"Running"* && ${line} != *"Completed"* ]]; then
+      ready="false"
+    fi
+  done
+  if [[  "${ready}" = "true" ]]; then
+    return 0
+  fi
+
+  echo "${pods_str}"
+  return 1
+}
+
+waitForPodsReady() {
+  echo "Waiting for pods to be ready in ${1}..."
+  withRetriesMaxTime 600 10 _waitForPodsReady "${1}"
+  echo "All pods ready."
+}
+
+waitForJob() {
+  echo "Waiting for job ${1} to complete..."
+  local start_time=${SECONDS}
+  until kubectl get jobs -n "${2}" "${1}" -o jsonpath='{.status.conditions[?(@.type=="Complete")].status}' | grep True ; do
+      sleep 1 ;
+  done
+  run_time=0
+  (( run_time = SECONDS - start_time ))
+  echo "Job ${1} ran for ${run_time} seconds."
+}
+
+checkDeploymentRolledOut() {
+  local ns="$1"
+  local name="$2"
+  
+  local total_replicas
+  local ready
+  local updated
+  local available
+
+  total_replicas=$(kubectl get deployment "${name}" -n "${ns}" -o jsonpath='{.spec.replicas}')
+  ready=$(kubectl get deployment "${name}" -n "${ns}" -o jsonpath='{.status.readyReplicas}')
+  updated=$(kubectl get deployment "${name}" -n "${ns}" -o jsonpath='{.status.updatedReplicas}')
+  available=$(kubectl get deployment "${name}" -n "${ns}" -o jsonpath='{.status.availableReplicas}')
+
+  echo "ready=${ready}, updated=${updated}, available=${available}"
+  if ((updated == total_replicas && available == total_replicas)); then
+    echo "rollout complete"
+    return 0
+  fi
+  return 1
+}
+# Return 1 if the specific error code percentage exceed corresponding threshold
+errorPercentBelow() {
+  local LOG=${1}
+  local ERR_CODE=${2}
+  local LIMIT=${3}
+  local s
+  s=$(grep "Code ${ERR_CODE}" "${LOG}")
+  local regex="Code ${ERR_CODE} : [0-9]+ \\(([0-9]+)\\.[0-9]+ %\\)"
+  if [[ ${s} =~ ${regex} ]]; then
+    local pctErr="${BASH_REMATCH[1]}"
+    if (( pctErr > LIMIT )); then
+      return 1
+    fi
+    echo "Errors percentage is within threshold"
+  fi
+  return 0
+}
+
+# Make a copy of test manifests in case either to/from branch doesn't contain them.
+copy_test_files() {
+  rm -Rf "${TMP_DIR}"
+  mkdir -p "${TMP_DIR}"
+  echo "${WD}"
+  cp -f "${WD}"/templates/* "${TMP_DIR}"/.
+}
+
+resetCluster() {
+  echo "Cleaning cluster by removing namespaces ${ISTIO_NAMESPACE} and ${TEST_NAMESPACE}"
+  deleteWithWait namespace "${ISTIO_NAMESPACE}"
+  deleteWithWait namespace "${TEST_NAMESPACE}"
+  echo "All namespaces deleted. Recreating ${ISTIO_NAMESPACE} and ${TEST_NAMESPACE}"
```

Comment:
i don't think this is good enough as a reset. CRDs and cluster resources must be cleaned up also, see https://istio.io/latest/docs/setup/install/istioctl/#uninstall-istio

Reply:
Agreed. I'll fix it.